Errors caught in middleware error handler cause process crash

<!-- Thanks for reporting an issue! Please fill out the blanks below. -->

## What are the steps to reproduce this issue?
1. Write code that configures HoneyBadger (v4.3.1), making sure not to disable breadcrumbs: 

```
Honeybadger.configure({
    apiKey: process.env.HONEYBADGER_API_KEY,
    environment,
    enableUncaught: true,
    enableUnhandledRejection: true,
    breadcrumbsEnabled: true,
    debug: false,
    revision: ver,
  });
```

3. Use the HB middleware as the last in the stack:
```
 app.use(Honeybadger.errorHandler);
``` 

## What happens?
Calling `Honeybadger.notify()` works as expected. 
However, when errors occur in Express controllers, there is an error in the HB client library that in turn causes an uncaught exception and hence process termination.

## What were you expecting to happen?
HB does not cause process to crash and successfully sends along breadcrumbs for errors caught in the middleware

## Any logs, error output, etc?
```
[2022-10-03T18:30:38.701Z] [ERROR] Invalid token. {"request":{"path":"/api/jobs","method":"POST","query":{},"params":{},"startTime":"Mon Oct 03 2022 15:30:38 GMT-0300 (Chile Summer Time)"}}
[2022-10-03T18:30:38.707Z] [ERROR] TypeError: Cannot read property 'breadcrumbs' of undefined
    at Honeybadger.Client.addBreadcrumb (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:991:33)
    at Honeybadger.Client.notify (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:847:14)
    at Honeybadger.errorHandler (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:1447:10)
    at Layer.handle_error (/our-app/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/our-app/node_modules/express/lib/router/index.js:315:13)
    at/our-app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/our-app/node_modules/express/lib/router/index.js:335:12)
    at next (/our-app/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle_error (/our-app/node_modules/express/lib/router/layer.js:73:5)
    at trim_prefix (/our-app/node_modules/express/lib/router/index.js:315:13)
    at/our-app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/our-app/node_modules/express/lib/router/index.js:335:12)
    at next (/our-app/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle_error (/our-app/node_modules/express/lib/router/layer.js:67:12)
    at trim_prefix (/our-app/node_modules/express/lib/router/index.js:315:13)
    at/our-app/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/our-app/node_modules/express/lib/router/index.js:335:12)
    at next (/our-app/node_modules/express/lib/router/index.js:275:10)
    at Strategy.strategy.error (/our-app/node_modules/passport/lib/middleware/authenticate.js:343:9)
    at verified (/our-app/node_modules/passport-http-bearer/lib/strategy.js:118:28)
    at/our-app/config/authentication.js:79:5
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {"source":"console","stacktrace":"Stacktrace\n    at Function.logerror (/our-app/node_modules/express/lib/application.js:630:43)\n    at processImmediate (internal/timers.js:463:21)\n    at process.topLevelDomainCallback (domain.js:138:15)\n    at process.callbackTrampoline (internal/async_hooks.js:124:14)"}
/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:991
        var breadcrumbs = store.breadcrumbs;
                                ^

TypeError: Cannot read property 'breadcrumbs' of undefined
    at Honeybadger.Client.addBreadcrumb (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:991:33)
    at Honeybadger.Client.notify (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:847:14)
    at process.honeybadgerUncaughtExceptionListener (/our-app/node_modules/@honeybadger-io/js/dist/server/honeybadger.js:1383:28)
    at process.emit (events.js:326:22)
    at process.EventEmitter.emit (domain.js:483:12)
    at process._fatalException (internal/process/execution.js:165:25)

Process finished with exit code 7

``` 

## Any other comments?
We worked around the issue for now turning off breadcrumbs:
```
Honeybadger.configure({
    ...
    breadcrumbsEnabled: false,
    ...
  });
```


## What versions are you using?
**Operating System:** …
MacOS (local env) and Ubuntu 20.04 (Server)

**Package Name:** …
**Package Version:** …
4.3.1


<img src="https://front.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_c9rkx)

--- a/packages/core/src/client.ts
+++ b/packages/core/src/client.ts
@@ -22,9 +22,7 @@ import {
   Notice,
   Noticeable,
   HoneybadgerStore,
-  BacktraceFrame,
-  BreadcrumbRecord,
-  DefaultStoreContents, Transport, NoticeTransportPayload
+  BacktraceFrame, Transport, NoticeTransportPayload
 } from './types'
 import { GlobalStore } from './store';
 
@@ -43,7 +41,7 @@ const NOT_BLANK = /\S/
 export abstract class Client {
   protected __pluginsExecuted = false
 
-  protected __store: HoneybadgerStore<{ context: Record<string, unknown>; breadcrumbs: BreadcrumbRecord[] }> = null;
+  protected __store: HoneybadgerStore = null;
   protected __beforeNotifyHandlers: BeforeNotifyHandler[] = []
   protected __afterNotifyHandlers: AfterNotifyHandler[] = []
   protected __getSourceFileHandler: (path: string) => Promise<string>
@@ -80,9 +78,7 @@ export abstract class Client {
       ...opts,
     }
 
-    // First, we go with the global (shared) store.
-    // Webserver middleware can then switch to the AsyncStore for async context tracking.
-    this.__store = new GlobalStore({ context: {}, breadcrumbs: [] })
+    this.__initStore();
     this.__transport = transport
     this.logger = logger(this)
   }
@@ -107,8 +103,8 @@ export abstract class Client {
     return this
   }
 
-  __setStore(store: HoneybadgerStore<{ context: Record<string, unknown>; breadcrumbs: BreadcrumbRecord[] }>) {
-    this.__store = store
+  protected __initStore() {
+    this.__store = new GlobalStore({ context: {}, breadcrumbs: [] }, this.config.maxBreadcrumbs);
   }
 
   beforeNotify(handler: BeforeNotifyHandler): Client {
@@ -122,31 +118,24 @@ export abstract class Client {
   }
 
   setContext(context: Record<string, unknown>): Client {
-    if (typeof context === 'object') {
-      const store = this.__store.getStore()
-      store.context = merge(store.context, context)
+    if (typeof context === 'object' && context != null) {
+      this.__store.setContext(context)
     }
     return this
   }
 
   resetContext(context?: Record<string, unknown>): Client {
     this.logger.warn('Deprecation warning: `Honeybadger.resetContext()` has been deprecated; please use `Honeybadger.clear()` instead.')
-    const store = this.__store.getStore()
-
+    this.__store.clear()
     if (typeof context === 'object' && context !== null) {
-      store.context = context
-    }
-    else {
-      store.context = {}
+      this.__store.setContext(context)
     }
 
     return this
   }
 
   clear(): Client {
-    const store = this.__store.getStore()
-    store.context = {}
-    store.breadcrumbs = []
+    this.__store.clear()
 
     return this
   }
@@ -196,8 +185,8 @@ export abstract class Client {
       }
     })
 
-    const breadcrumbs = this.__getStoreContentsOrDefault().breadcrumbs
-    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? breadcrumbs.slice() : []
+    const breadcrumbs = this.__store.getContents('breadcrumbs')
+    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? breadcrumbs : []
 
     getSourceForBacktrace(sourceCodeData, this.__getSourceFileHandler)
       .then(sourcePerTrace => {
@@ -304,7 +293,7 @@ export abstract class Client {
       return null
     }
 
-    const context = this.__getStoreContentsOrDefault().context
+    const context = this.__store.getContents('context')
     const noticeTags = this.__constructTags(notice.tags)
     const contextTags = this.__constructTags(context['tags'])
     const configTags = this.__constructTags(this.config.tags)
@@ -346,22 +335,22 @@ export abstract class Client {
     const category = opts.category || 'custom'
     const timestamp = new Date().toISOString()
 
-    const store = this.__store.getStore()
-    let breadcrumbs = store.breadcrumbs
-    breadcrumbs.push({
+    this.__store.addBreadcrumb({
       category: category as string,
-      message: message,
+      message,
       metadata: metadata as Record<string, unknown>,
-      timestamp: timestamp
+      timestamp
     })
 
-    const limit = this.config.maxBreadcrumbs
-    if (breadcrumbs.length > limit) {
-      breadcrumbs = breadcrumbs.slice(breadcrumbs.length - limit)
-    }
-    store.breadcrumbs = breadcrumbs
-
     return this
+  }
+
+  __getBreadcrumbs() {
+    return this.__store.getContents('breadcrumbs').slice()
+  }
+
+  __getContext() {
+    return this.__store.getContents('context')
   }
 
   protected __developmentMode(): boolean {
@@ -416,20 +405,5 @@ export abstract class Client {
     }
 
     return tags.toString().split(TAG_SEPARATOR).filter((tag) => NOT_BLANK.test(tag))
-  }
-
-  /**
-   * For ALS, the store may be uninitialized (if .run()` has not been called).
-   * This provides an easy way to read the existing store object or fall back to a default.
-   * Returns *a copy* of the s tore contents.
-   */
-  protected __getStoreContentsOrDefault(): DefaultStoreContents {
-    const existingStoreContents = this.__store.getStore();
-    const storeContents = existingStoreContents || {};
-    return {
-      context: {},
-      breadcrumbs: [],
-      ...storeContents
-    };
   }
 }

--- a/packages/core/src/index.ts
+++ b/packages/core/src/index.ts
@@ -1,4 +1,4 @@
 export { Client } from './client'
-export * as Store from './store'
+export * from './store'
 export * as Types from './types'
 export * as Util from './util'

--- a/packages/core/src/store.ts
+++ b/packages/core/src/store.ts
@@ -1,18 +1,45 @@
-import { HoneybadgerStore } from './types';
+import { HoneybadgerStore, StoreContents } from './types';
+import { merge } from './util';
 
-export class GlobalStore<T> implements HoneybadgerStore<T> {
-  private store: T
+export class GlobalStore implements HoneybadgerStore {
+  private readonly contents: StoreContents;
+  private readonly breadcrumbsLimit: number;
 
-  constructor(store: T) {
-    this.store = store;
+  constructor(contents: StoreContents, breadcrumbsLimit: number) {
+    this.contents = contents;
+    this.breadcrumbsLimit = breadcrumbsLimit;
   }
 
-  getStore(): T {
-    return this.store
+  static create(contents: StoreContents, breadcrumbsLimit: number) {
+    return new GlobalStore(contents, breadcrumbsLimit);
   }
 
-  run<R, TArgs extends unknown[]>(store: T, callback: (...args: TArgs) => R, ...args: TArgs): R {
-    this.store = store;
-    return callback(...args);
+  available() {
+    return true
+  }
+
+  getContents(key?: keyof StoreContents) {
+    const value = key ? this.contents[key] : this.contents;
+    return JSON.parse(JSON.stringify(value));
+  }
+
+  setContext(context) {
+    this.contents.context = merge(this.contents.context, context || {});
+  }
+
+  addBreadcrumb(breadcrumb) {
+    if (this.contents.breadcrumbs.length == this.breadcrumbsLimit) {
+      this.contents.breadcrumbs.shift()
+    }
+    this.contents.breadcrumbs.push(breadcrumb);
+  }
+
+  clear() {
+    this.contents.context = {}
+    this.contents.breadcrumbs = []
+  }
+
+  run(callback) {
+    return callback();
   }
 }

--- a/packages/core/src/types.ts
+++ b/packages/core/src/types.ts
@@ -1,5 +1,4 @@
 import { Client } from './client'
-import type { AsyncLocalStorage } from 'async_hooks';
 
 export interface Logger {
   log(...args: unknown[]): unknown
@@ -163,5 +162,29 @@ export interface Transport {
   send(options: TransportOptions, payload?: NoticeTransportPayload | undefined): Promise<{ statusCode: number; body: string; }>
 }
 
-export type HoneybadgerStore<T> = Pick<AsyncLocalStorage<T>, 'getStore' | 'run'>
-export type DefaultStoreContents = {context: Record<string, unknown>, breadcrumbs: BreadcrumbRecord[]}
+export interface HoneybadgerStore {
+  /**
+   * Is the store available for writing and reading?
+   */
+  available(): boolean
+
+  /**
+   * With no arguments, returns a copy of the store contents for reading.
+   * When a key is supplied, returns the value of that key in the store.
+   */
+  getContents(): StoreContents
+  getContents<K extends keyof StoreContents>(key: K): StoreContents[K]
+
+  setContext(context: Record<string, unknown>): void
+
+  addBreadcrumb(breadcrumb: BreadcrumbRecord): void
+
+  clear(): void
+
+  run<R>(callback: () => R, ...opts: unknown[]): R
+}
+
+export type StoreContents = {
+  context: Record<string, unknown>,
+  breadcrumbs: BreadcrumbRecord[]
+}

--- a/packages/core/test/client.test.ts
+++ b/packages/core/test/client.test.ts
@@ -47,7 +47,7 @@ describe('client', function () {
   })
 
   it('has a context object', function () {
-    expect(client.getContext()).toEqual({})
+    expect(client.__getContext()).toEqual({})
   })
 
   describe('setContext', function () {
@@ -58,8 +58,7 @@ describe('client', function () {
         foo: 'bar'
       })
 
-      expect(client.getContext().user_id).toEqual('1')
-      expect(client.getContext().foo).toEqual('bar')
+      expect(client.__getContext()).toEqual({ user_id: '1', foo: 'bar' })
     })
 
     it('is chainable', function () {
@@ -70,7 +69,7 @@ describe('client', function () {
 
     it('does not accept non-objects', function () {
       client.setContext(<never>'foo')
-      expect(client.getContext()).toEqual({})
+      expect(client.__getContext()).toEqual({})
     })
 
     it('keeps previous context when called with non-object', function () {
@@ -78,7 +77,7 @@ describe('client', function () {
         foo: 'bar'
       }).setContext(<never>false)
 
-      expect(client.getContext()).toEqual({
+      expect(client.__getContext()).toEqual({
         foo: 'bar'
       })
     })
@@ -91,13 +90,13 @@ describe('client', function () {
         user_id: '1'
       })
 
-      expect(client.getContext()).not.toEqual({})
-      expect(client.getBreadcrumbs()).not.toEqual([])
+      expect(client.__getContext()).not.toEqual({})
+      expect(client.__getBreadcrumbs()).not.toEqual([])
 
       client.clear()
 
-      expect(client.getContext()).toEqual({})
-      expect(client.getBreadcrumbs()).toEqual([])
+      expect(client.__getContext()).toEqual({})
+      expect(client.__getBreadcrumbs()).toEqual([])
     })
   })
 
@@ -107,7 +106,7 @@ describe('client', function () {
         user_id: '1'
       }).resetContext()
 
-      expect(client.getContext()).toEqual({})
+      expect(client.__getContext()).toEqual({})
     })
 
     it('replaces the context with arguments', function () {
@@ -117,7 +116,7 @@ describe('client', function () {
         foo: 'bar'
       })
 
-      expect(client.getContext()).toEqual({
+      expect(client.__getContext()).toEqual({
         foo: 'bar'
       })
     })
@@ -127,7 +126,7 @@ describe('client', function () {
         foo: 'bar'
       }).resetContext(<never>'foo')
 
-      expect(client.getContext()).toEqual({})
+      expect(client.__getContext()).toEqual({})
     })
 
     it('is chainable', function () {
@@ -706,15 +705,15 @@ describe('client', function () {
 
   describe('addBreadcrumb', function () {
     it('has default breadcrumbs', function () {
-      expect(client.getBreadcrumbs()).toEqual([])
+      expect(client.__getBreadcrumbs()).toEqual([])
     })
 
     it('adds a breadcrumb with defaults', function () {
       client.addBreadcrumb('expected message')
 
-      expect(client.getBreadcrumbs().length).toEqual(1)
+      expect(client.__getBreadcrumbs().length).toEqual(1)
 
-      const crumb = client.getBreadcrumbs()[0]
+      const crumb = client.__getBreadcrumbs()[0]
 
       expect(crumb.message).toEqual('expected message')
       expect(crumb.category).toEqual('custom')
@@ -727,8 +726,8 @@ describe('client', function () {
         category: 'test'
       })
 
-      expect(client.getBreadcrumbs().length).toEqual(1)
-      expect(client.getBreadcrumbs()[0].category).toEqual('test')
+      expect(client.__getBreadcrumbs().length).toEqual(1)
+      expect(client.__getBreadcrumbs()[0].category).toEqual('test')
     })
 
     it('overrides the default metadata', function () {
@@ -738,8 +737,8 @@ describe('client', function () {
         }
       })
 
-      expect(client.getBreadcrumbs().length).toEqual(1)
-      expect(client.getBreadcrumbs()[0].metadata).toEqual({
+      expect(client.__getBreadcrumbs().length).toEqual(1)
+      expect(client.__getBreadcrumbs()[0].metadata).toEqual({
         key: 'expected value'
       })
     })
@@ -755,9 +754,9 @@ describe('client', function () {
         metadata: metadata
       })
 
-      expect(client.getBreadcrumbs().length).toEqual(2)
-      expect(client.getBreadcrumbs()[0].metadata).toEqual(client.getBreadcrumbs()[1].metadata)
-      expect(client.getBreadcrumbs()[0].metadata).not.toBe(client.getBreadcrumbs()[1].metadata)
+      expect(client.__getBreadcrumbs().length).toEqual(2)
+      expect(client.__getBreadcrumbs()[0].metadata).toEqual(client.__getBreadcrumbs()[1].metadata)
+      expect(client.__getBreadcrumbs()[0].metadata).not.toBe(client.__getBreadcrumbs()[1].metadata)
     })
 
     it('maintains the size of the breadcrumbs queue', function () {
@@ -765,10 +764,10 @@ describe('client', function () {
         client.addBreadcrumb('expected message ' + i)
       }
 
-      expect(client.getBreadcrumbs().length).toEqual(40)
+      expect(client.__getBreadcrumbs().length).toEqual(40)
 
-      expect(client.getBreadcrumbs()[0].message).toEqual('expected message 6')
-      expect(client.getBreadcrumbs()[39].message).toEqual('expected message 45')
+      expect(client.__getBreadcrumbs()[0].message).toEqual('expected message 6')
+      expect(client.__getBreadcrumbs()[39].message).toEqual('expected message 45')
     })
 
     it('sends breadcrumbs by default', function () {

--- a/packages/core/test/helpers.ts
+++ b/packages/core/test/helpers.ts
@@ -31,14 +31,6 @@ export class TestClient extends BaseClient {
     super(opts, transport);
   }
 
-  public getContext() {
-    return this.__store.getStore().context
-  }
-
-  public getBreadcrumbs() {
-    return this.__store.getStore().breadcrumbs
-  }
-
   public getPayload(noticeable: Noticeable, name: string | Partial<Notice> = undefined, extra: Partial<Notice> = undefined) {
     // called in client.notify()
     const notice = this.makeNotice(noticeable, name, extra)
@@ -52,7 +44,7 @@ export class TestClient extends BaseClient {
       }
     })
 
-    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? this.getBreadcrumbs().slice() : []
+    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? this.__getBreadcrumbs() : []
 
     // called in (server|browser).__send()
     return this.__buildPayload(notice)

--- a/packages/js/src/server.ts
+++ b/packages/js/src/server.ts
@@ -9,12 +9,11 @@ import uncaughtException from './server/integrations/uncaught_exception'
 import unhandledRejection from './server/integrations/unhandled_rejection'
 import { errorHandler, requestHandler } from './server/middleware'
 import { lambdaHandler } from './server/aws_lambda'
-import { AsyncStore } from './server/async_store'
 import { ServerTransport } from './server/transport';
+import { StackedStore } from './server/stacked_store';
 
 const { endpoint } = Util
 
-const kHoneybadgerStore = Symbol.for('kHoneybadgerStore');
 class Honeybadger extends Client {
   /** @internal */
   protected __beforeNotifyHandlers: Types.BeforeNotifyHandler[] = [
@@ -65,6 +64,11 @@ class Honeybadger extends Client {
     return super.configure(opts)
   }
 
+  protected __initStore() {
+    // @ts-ignore
+    this.__store = new StackedStore(this.config.maxBreadcrumbs);
+  }
+
   checkIn(id: string): Promise<void> {
     return this.__transport
       .send({
@@ -87,18 +91,13 @@ class Honeybadger extends Client {
   // 1. Using AsyncLocalStorage so the context is tracked across async operations.
   // 2. Attaching the store contents to the request object,
   //   so, even if the store is destroyed, we can still recover the context for a given request
+  //   (for example, in an error-handling middleware)
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   public withRequest<R>(
     request: Record<symbol, unknown>,
     handler: (...args: never[]) => R,
     onError?: (...args: unknown[]) => unknown
   ): R|void {
-    const storeObject = (request[kHoneybadgerStore] || this.__getStoreContentsOrDefault()) as Types.DefaultStoreContents;
-    this.__setStore(AsyncStore);
-    if (!request[kHoneybadgerStore]) {
-      request[kHoneybadgerStore] = storeObject;
-    }
-
     if (onError) {
       // ALS is fine for context-tracking, but `domain` allows us to catch errors
       // thrown asynchronously (timers, event emitters)
@@ -107,12 +106,16 @@ class Honeybadger extends Client {
       // Note that this doesn't still handle all cases. `domain` has its own problems:
       // See https://github.com/honeybadger-io/honeybadger-js/pull/711
       const dom = domain.create();
-      const onErrorWithContext = (err) => this.__store.run(storeObject, () => onError(err));
+      const onErrorWithContext = (err) => this.__store.run(() => onError(err), request);
       dom.on('error', onErrorWithContext);
       handler = dom.bind(handler);
     }
 
-    return this.__store.run(storeObject, handler);
+    return this.__store.run(handler, request);
+  }
+
+  public run<R>(callback: (...args: never[]) => R): R {
+    return this.__store.run(callback);
   }
 }
 

--- a/packages/js/src/server/async_store.ts
+++ b/packages/js/src/server/async_store.ts
@@ -1,13 +1,86 @@
-import { Types, Store as StoreModule } from '@honeybadger-io/core'
+import { Types } from '@honeybadger-io/core';
+import { AsyncLocalStorage } from 'async_hooks';
 
-let Store: Types.HoneybadgerStore<{ context: Record<string, unknown>; breadcrumbs: Types.BreadcrumbRecord[]; }>
-try {
-  // eslint-disable-next-line @typescript-eslint/no-var-requires
-  const { AsyncLocalStorage } = require('async_hooks')
-  Store = new AsyncLocalStorage()
-}
-catch (e) {
-  Store = new StoreModule.GlobalStore({ context: {}, breadcrumbs: [] })
-}
+const kHoneybadgerStore = Symbol.for('kHoneybadgerStore');
 
-export const AsyncStore = Store
+export class AsyncStore implements Types.HoneybadgerStore {
+  private als: AsyncLocalStorage<Types.StoreContents>;
+  private readonly contents: Types.StoreContents;
+  private readonly breadcrumbsLimit: number;
+
+  constructor(
+    asyncLocalStorage: AsyncLocalStorage<Types.StoreContents>,
+    contents: Types.StoreContents, breadcrumbsLimit: number
+  ) {
+    this.als = asyncLocalStorage;
+    this.contents = contents;
+    this.breadcrumbsLimit = breadcrumbsLimit;
+  }
+
+  /**
+   * Attempt to create a new AsyncStore instance
+   */
+  static create(contents: Types.StoreContents, breadcrumbsLimit: number): AsyncStore|null {
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-var-requires
+      const { AsyncLocalStorage } = require('async_hooks');
+      return new AsyncStore(new AsyncLocalStorage, contents, breadcrumbsLimit);
+    } catch (e) {
+      return null
+    }
+  }
+
+  /**
+   * This returns the live store object, so we can mutate it.
+   * If we're in an async context (a `run()` callback), the stored contents at this point will be returned.
+   * Otherwise, the initial stored contents will be returned.
+   */
+  __currentContents() {
+    return this.als.getStore() || this.contents
+  }
+
+  getContents(key?: keyof Types.StoreContents) {
+    const value = key ? this.__currentContents()[key] : this.__currentContents();
+    return JSON.parse(JSON.stringify(value));
+  }
+
+  available() {
+    return !!this.als.getStore()
+  }
+
+  setContext(context: Record<string, unknown>): void {
+    Object.assign(this.__currentContents().context, context || {});
+  }
+
+  addBreadcrumb(breadcrumb) {
+    if (this.__currentContents().breadcrumbs.length == this.breadcrumbsLimit) {
+      this.__currentContents().breadcrumbs.shift()
+    }
+    this.__currentContents().breadcrumbs.push(breadcrumb);
+  }
+
+  clear() {
+    this.__currentContents().context = {}
+    this.__currentContents().breadcrumbs = []
+  }
+
+  run<R>(callback: () => R, request?: Record<symbol, unknown>): R {
+    // When entering an async context, we pass in *a copy* of the initial contents.
+    // This allows every request to start from the same state
+    // and add data specific to them, without overlapping.
+
+    if (!request) {
+      return this.als.run(this.getContents(), callback);
+    }
+
+    let contents: Types.StoreContents;
+    if (request[kHoneybadgerStore]) {
+      contents = request[kHoneybadgerStore] as Types.StoreContents
+    } else {
+      contents = this.getContents();
+      request[kHoneybadgerStore] = contents;
+    }
+
+    return this.als.run(contents, callback);
+  }
+}

--- a/packages/js/src/server/aws_lambda.ts
+++ b/packages/js/src/server/aws_lambda.ts
@@ -1,7 +1,7 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
-import { Client as Honeybadger, Types } from '@honeybadger-io/core'
+import { Types } from '@honeybadger-io/core'
+import Honeybadger from '../server'
 import type { Handler, Callback, Context } from 'aws-lambda'
-import { AsyncStore } from './async_store';
 
 export type SyncHandler<TEvent = any, TResult = any> = (
     event: TEvent,
@@ -18,7 +18,7 @@ function isHandlerSync(handler: Handler): handler is SyncHandler {
   return handler.length > 2
 }
 
-function reportToHoneybadger(hb: Honeybadger, err: Error | string | null, callback: (err: Error | string | null) => void) {
+function reportToHoneybadger(hb: typeof Honeybadger, err: Error | string | null, callback: (err: Error | string | null) => void) {
   hb.notify(err, {
     afterNotify: function () {
       hb.clear()
@@ -27,11 +27,10 @@ function reportToHoneybadger(hb: Honeybadger, err: Error | string | null, callba
   })
 }
 
-function asyncHandler<TEvent = any, TResult = any>(handler: AsyncHandler<TEvent, TResult>, hb: Honeybadger): AsyncHandler<TEvent, TResult> {
+function asyncHandler<TEvent = any, TResult = any>(handler: AsyncHandler<TEvent, TResult>, hb: typeof Honeybadger): AsyncHandler<TEvent, TResult> {
   return function wrappedLambdaHandler(event, context) {
-    hb.__setStore(AsyncStore)
     return new Promise<TResult>((resolve, reject) => {
-      AsyncStore.run({ context: {}, breadcrumbs: [] }, () => {
+      hb.run(() => {
         const timeoutHandler = setupTimeoutWarning(hb, context)
         try {
           const result = handler(event, context);
@@ -55,10 +54,9 @@ function asyncHandler<TEvent = any, TResult = any>(handler: AsyncHandler<TEvent,
   }
 }
 
-function syncHandler<TEvent = any, TResult = any>(handler: SyncHandler<TEvent, TResult>, hb: Honeybadger): SyncHandler<TEvent, TResult> {
+function syncHandler<TEvent = any, TResult = any>(handler: SyncHandler<TEvent, TResult>, hb: typeof Honeybadger): SyncHandler<TEvent, TResult> {
   return function wrappedLambdaHandler(event, context, cb) {
-    hb.__setStore(AsyncStore)
-    AsyncStore.run({ context: {}, breadcrumbs: [] }, () => {
+    hb.run(() => {
       const timeoutHandler = setupTimeoutWarning(hb, context)
       try {
         handler(event, context, (error, result) => {
@@ -77,11 +75,11 @@ function syncHandler<TEvent = any, TResult = any>(handler: SyncHandler<TEvent, T
   }
 }
 
-function shouldReportTimeoutWarning(hb: Honeybadger, context: Context): boolean {
+function shouldReportTimeoutWarning(hb: typeof Honeybadger, context: Context): boolean {
   return typeof context.getRemainingTimeInMillis === 'function' && !!((hb.config as Types.ServerlessConfig).reportTimeoutWarning)
 }
 
-function setupTimeoutWarning(hb: Honeybadger, context: Context) {
+function setupTimeoutWarning(hb: typeof Honeybadger, context: Context) {
   if (!shouldReportTimeoutWarning(hb, context)) {
     return
   }
@@ -95,7 +93,7 @@ function setupTimeoutWarning(hb: Honeybadger, context: Context) {
 
 export function lambdaHandler<TEvent = any, TResult = any>(handler: Handler<TEvent, TResult>): Handler<TEvent, TResult> {
   // eslint-disable-next-line @typescript-eslint/no-this-alias
-  const hb: Honeybadger = this
+  const hb: typeof Honeybadger = this
   if (isHandlerSync(handler)) {
     return syncHandler(handler, hb)
   }

--- a/packages/js/src/server/stacked_store.ts
+++ b/packages/js/src/server/stacked_store.ts
@@ -1,0 +1,51 @@
+import { GlobalStore, Types } from '@honeybadger-io/core';
+import { AsyncStore } from './async_store';
+
+/**
+ * A store that's really a "stack" of stores (async and global)
+ * Will proxy calls to the async store if it's active (ie when we enter an async context via `ALS.run()`),
+ * otherwise it will fall back to the global store.
+ * Both stores in the stack start out with the same contents object.
+ * Note that the asyncStore may be null if ALS is not supported.
+ */
+export class StackedStore implements Types.HoneybadgerStore {
+  private readonly contents: Types.StoreContents
+  private readonly asyncStore: AsyncStore|null
+  private readonly globalStore: GlobalStore
+
+  constructor(breadcrumbsLimit: number) {
+    this.contents = { context: {}, breadcrumbs: [] };
+    this.asyncStore = AsyncStore.create(this.contents, breadcrumbsLimit);
+    this.globalStore = GlobalStore.create(this.contents, breadcrumbsLimit);
+  }
+
+  __activeStore() {
+    return this.asyncStore?.available() ? this.asyncStore : this.globalStore
+  }
+
+  available() {
+    return true;
+  }
+
+  // @ts-ignore
+  getContents(key?: keyof Types.StoreContents) {
+    return this.__activeStore().getContents(key);
+  }
+
+  setContext(context) {
+    this.__activeStore().setContext(context);
+  }
+
+  addBreadcrumb(breadcrumb) {
+    this.__activeStore().addBreadcrumb(breadcrumb)
+  }
+
+  clear() {
+    this.__activeStore().clear();
+  }
+
+  run<R>(callback: () => R, request?: Record<symbol, unknown>): R {
+    // We explicitly favour the async store here, if ALS is supported. It's the whole point of the `run()` method
+    return this.asyncStore ? this.asyncStore.run(callback, request) : this.globalStore.run(callback);
+  }
+}

--- a/packages/js/test/unit/helpers.ts
+++ b/packages/js/test/unit/helpers.ts
@@ -31,14 +31,6 @@ export class TestClient extends BaseClient {
     super(opts, transport);
   }
 
-  public getContext() {
-    return this.__store.getStore().context
-  }
-
-  public getBreadcrumbs() {
-    return this.__store.getStore().breadcrumbs
-  }
-
   public getPayload(noticeable: Types.Noticeable, name: string | Partial<Types.Notice> = undefined, extra: Partial<Types.Notice> = undefined) {
     // called in client.notify()
     const notice = this.makeNotice(noticeable, name, extra)
@@ -52,7 +44,7 @@ export class TestClient extends BaseClient {
       }
     })
 
-    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? this.getBreadcrumbs().slice() : []
+    notice.__breadcrumbs = this.config.breadcrumbsEnabled ? this.__getBreadcrumbs() : []
 
     // called in (server|browser).__send()
     return this.__buildPayload(notice)

--- a/packages/js/test/unit/server.test.ts
+++ b/packages/js/test/unit/server.test.ts
@@ -88,6 +88,44 @@ describe('server client', function () {
     })
   })
 
+  it('combines previous global state when reporting', function () {
+    let expectedAssertions = 2; // Safeguard to ensure all handlers are called
+
+    client.addBreadcrumb('global 1')
+    client.addBreadcrumb('global 2')
+    const req1 = {}
+    const req2 = {}
+    client.withRequest(req1, () => {
+      client.addBreadcrumb('async 1 from request 1')
+    })
+    client.withRequest(req2, () => {
+      client.addBreadcrumb('async 1 from request 2')
+      expect(client.__getBreadcrumbs()).toHaveLength(3)
+      expect(client.__getBreadcrumbs().map(({ message }) => message)).toEqual(
+        ['global 1', 'global 2', 'async 1 from request 2']
+      )
+      expectedAssertions--
+    })
+    client.withRequest(req1, () => {
+      client.addBreadcrumb('async 2 from request 1')
+      expect(client.__getBreadcrumbs()).toHaveLength(4)
+      expect(client.__getBreadcrumbs().map(({ message }) => message)).toEqual(
+        ['global 1', 'global 2', 'async 1 from request 1', 'async 2 from request 1']
+      )
+      expectedAssertions--
+    })
+
+    client.addBreadcrumb('global 3')
+    expect(client.__getBreadcrumbs()).toHaveLength(3)
+    expect(client.__getBreadcrumbs().map(({ message }) => message)).toEqual(
+      ['global 1', 'global 2', 'global 3']
+    )
+
+    if (expectedAssertions !== 0) {
+      throw new Error(`Not all assertions ran. ${expectedAssertions} assertions did not run.`)
+    }
+  })
+
   describe('afterNotify', function () {
     beforeEach(function () {
       client.configure({
@@ -207,7 +245,7 @@ describe('server client', function () {
       const errorHandler = (e) => {
         err = e;
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        context = (<any>client).__store.getStore().context;
+        context = (<any>client).__getContext();
       }
 
       client.withRequest({}, () => client.setContext({ a: true }), errorHandler)
@@ -232,12 +270,12 @@ describe('server client', function () {
       });
       client.withRequest(request, () => {
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        expect((<any>client).__store.getStore().context).toStrictEqual({ request1: true });
+        expect((<any>client).__getContext()).toStrictEqual({ request1: true });
       });
       client.withRequest(request, () => {
         setTimeout(() => {
           // eslint-disable-next-line @typescript-eslint/no-explicit-any
-          expect((<any>client).__store.getStore().context).toStrictEqual({ request1: true });
+          expect((<any>client).__getContext()).toStrictEqual({ request1: true });
           done();
         }, 200);
       });

--- a/packages/js/test/unit/server/middleware.server.test.ts
+++ b/packages/js/test/unit/server/middleware.server.test.ts
@@ -97,17 +97,15 @@ describe('Express Middleware', function () {
     app.use(client.requestHandler)
 
     app.get('/:reqId', (req, res) => {
-      const initialContext = client.__store.getStore().context;
+      const initialContext = client.__getContext();
       client.setContext({ reqId: req.params.reqId });
       setTimeout(() => {
         res.json({
           initial: initialContext,
-          final: client.__store.getStore().context
+          final: client.__getContext()
         });
-      }, 1000);
+      }, 500);
     });
-
-    app.use(client.errorHandler)
 
     return Promise.all([1, 2].map((i) => {
       return request(app).get(`/${i}`)
@@ -115,6 +113,39 @@ describe('Express Middleware', function () {
         .then((response) => {
           const expectedContexts = { initial: {}, final: { reqId: `${i}` } }
           expect(response.body).toStrictEqual(expectedContexts)
+        })
+    }));
+  })
+
+  it('preserves context in the error handlers', function() {
+    client.afterNotify((err, notice) => {
+      expect(notice.context.reqId).toEqual(notice.message)
+      expect(Object.keys(notice.context.initialContext)).toHaveLength(0)
+    })
+
+    const app = express()
+
+    app.use(client.requestHandler)
+
+    app.get('/:reqId', (req, _res) => {
+      client.setContext({ reqId: req.params.reqId, initialContext: client.__getContext() });
+      setTimeout(function asyncThrow() {
+        throw new Error(req.params.reqId)
+      }, 500)
+    });
+
+    app.use(client.errorHandler)
+
+    app.use(function(_err, _req, res, _next) {
+      res.json(client.__getContext(), 500);
+    })
+
+    return Promise.all([80, 90].map((i) => {
+      return request(app).get(`/${i}`)
+        .expect(500)
+        .then((response) => {
+          const expectedContext = { reqId: `${i}`, initialContext: {} }
+          expect(response.body).toStrictEqual(expectedContext)
         })
     }));
   })

